### PR TITLE
Deployment for Marlowe Dash

### PR DESCRIPTION
**Pull request by shmish111**
This PR builds a basic infrastructure for deploying Marlowe Dash to the web.

An example deployment is available at https://david.marlowe-dash.iohkdev.io/

Notes:
1. The terraform scripts have been updated to create new marlowe-dash EC2 instances and load balancer routes to them.
2. The serverless dream is over. Unfortunately I couldn’t work out a way to get AWS to use web sockets and serve static content from S3 for the same domain. This means for this new application we have had to go back to serving static files directly from the servant server. Maybe this is a sign of things to come, although the lambdas work well and save resources, the terraform code is very complex and they add the need for static builds. It might be time to revisit all of this. 
3. I discovered https://github.com/DBCDK/morph which we can use in place of nixops and which allows us to deploy from any machine that has ssh access to the EC2 instance. For now I have just added the new marlowe dash definitions but in another PR I will move the webghc machines over to morph as well as cleaning up the deployment README.md
4. I couldn’t quite work out how to integrate morph into the deployment scripts because it would involve nix-in-nix however I think there must be a way. I have added some very nasty hacking to make things easier (see ./deployment/default.nix line 125)
5. I have added ./deployment/morph/test.nix which enables us to test NixOS configurations for validity, this can be plumbed into Hydra in a future PR.

We’re closer to the dream of single command deployment but it’s not quite there yet. Perhaps reviewing this PR might spawn some more ideas…

Pre-submit checklist:
- Branch
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [ ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by shmish111** [comment]
nix-instantiate is failing saying there is an issue with the materialized view however I have updated it? I did notice that nixfmt changes it, could this be an issue?

**Comment by michaelpj** [comment]
> I did notice that nixfmt changes it, could this be an issue?

Yes. But it should be ignored. However I think sometimes it's bad at picking up the ignore instruction. If you can't get it to ignore it you can do `SKIP=nixpgs-fmt git commit ...`

**Comment by shmish111** [comment]
It seems I need to do git commit twice and then the `SKIP` is respected, weird

**Comment by shmish111** [comment]
I couldn't get it to work, I had to do ` --no-verify` which worked fine but isn't suitable for anything other than a single file commit of materialized view because it disables all checks

**Comment by shmish111** [comment]
oh weird, I've just noticed that the problem file is `nix/pkgs/haskell/materialized-unix/.plan.nix/marlowe-dashboard-server.nix` which is not in the nixfmt exclude setting. Since this is a file that I have added in this PR that suggests that I have done something wrong?

**Comment by michaelpj** [comment]
No, just looks like I screwed up the patterns somehow...

File: deployment/morph/overlays.nix
Comment (gilligan) [inline deployment/morph/overlays.nix:2]: Could you turn this comment into more of an action item? 
```
TODO: drop once <some link> is in the pinned nixpkgs
```

**Review by gilligan — CHANGES_REQUESTED** [review]
I only noticed some minor issues that i commented on which should be quick to resolve.

I do wonder a bit about what the plan is from here on out. This PR adds the use of `morph` to the deployment meaning we are now dealing with: nixops, terraform *and* morph. So this PR is essentially adding complexity.

I don't mind the use of morph per se but then then nixops should be removed from the deployment in general right (not necessarily as part of this PR) - Leaving a combination of terraform and morph, which is less overlapping than terraform and nixops (everything that terraform does could also be done with nixops itself).

**Comment by shmish111** [comment]
Sorry yes, I intend to get rid of NixOS, that's what I meant in my description about moving the webghc machines but I wasn't very clear.

File: deployment/default.nix
Comment (shmish111) [reply deployment/default.nix:80]: yes, I should have added a comment here. The quotes prevent wildcards from working in bash.

File: deployment/default.nix
Comment (shmish111) [reply deployment/default.nix:120]: added

File: deployment/morph/default-machine.nix
Comment (shmish111) [reply deployment/morph/default-machine.nix:28]: I have no idea. This file is a copy and paste x 2 from a file I created 3 years ago and it seems that hvm is considered better than pv these days so might as well.

File: deployment/morph/default-machine.nix
Comment (shmish111) [reply deployment/morph/default-machine.nix:33]: again, from old times. I am not a fan of the "safety" of rate-limiting journald over not losing any logs so I changed the default. Given it's the default machine you can override it when you want.

File: deployment/morph/overlays.nix
Comment (shmish111) [reply deployment/morph/overlays.nix:2]: yeah I think we don't need this any more, removing

File: deployment/morph/marlowe-dash.nix
Comment (shmish111) [reply deployment/morph/marlowe-dash.nix:32]: done.

File: deployment/morph/default-machine.nix
Comment (gilligan) [reply deployment/morph/default-machine.nix:28]: After some googling i landed here: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html

> For the best performance, we recommend that you use current generation instance types and HVM AMIs when you launch your instances.

So if `hvm` isn't the default in nixops you could either add this link or just state that amazon advises the use of `hvm` over `pv` now.

Comment (shmish111) [reply deployment/morph/default-machine.nix:29]: added

File: deployment/morph/default-machine.nix
Comment (gilligan) [reply deployment/morph/default-machine.nix:34]: The default is 100000 which is quite a lot of logging :sweat_smile: Given that - as you said yourself - this is the default configuration that can be overridden elsewhere i would probably just leave it to the default value. But that's up to you.

**Review by gilligan — APPROVED** [review]

**Pull request merged**
